File: crates/core/src/eval/ledger.rs
//! Where every byte of a file goes, grouped the way a reader asks about it.
//!
//! The report's "Where the bytes go" section and its full ledger are drawn
//! from this. One grouping rule and no judgment (see `docs/DESIGN-report-view.md`,
//! "Where the bytes go"):
//!
//! - **Part.** The top-level part a byte is in: the root's field it is under,
//!   or the root itself when the root is a list or a single value. A view
//!   that draws parts at a finer grain sums the rows under each of its own.
//! - **Group.** The variant of the nearest ancestor that is an element of a
//!   list: the case its switch took. MIDI events group as the kinds of event,
//!   SQLite pages as leaf and interior pages. Where the switch fell to its
//!   default, the element's own name says more than "bytes" does, so a PNG's
//!   `IDAT` chunks are `IDAT` and not the raw bytes every unknown chunk is.
//! - **Role.** `content`; `machinery`, which is every field another field's
//!   length, count, type or place depends on, and everything inside one;
//!   `padding`, a run sized to an alignment; `framing`, a JSON value's own
//!   braces and commas; and `gap`, bits no field covers.
//!
//! Gaps and padding are split into zero bytes and the rest, which takes
//! reading them. That is done after the walk, a step at a time, because the
//! walk must not read anything once it has started writing a node down (see
//! `watch.rs`).
//!
//! **Every byte once.** A field placed by an offset covers bits that the
//! structure around it may have left as a gap: an HDF5 root is ninety-six
//! bytes and every object after it is placed by an address, and a SQLite
//! page's cells are placed inside the stretch its header leaves over. The
//! kind totals live with that and hold the gap total down afterwards. The
//! ledger cannot, because it says which bytes are the gap, so every gap has
//! the fields placed over it taken out of it, except fields it is inside of:
//! a gap in a placed field's own structure is a gap that field left.

use std::collections::BTreeMap;

use super::*;

/// One line of the ledger.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LedgerRow {
    /// The top-level part, as a path from the walk's root.
    pub part: Vec<usize>,
    pub part_name: String,
    /// The variant these bytes are grouped under. Empty when no list element
    /// is above them, where the part is the only grouping there is.
    pub group: String,
    /// Where the group's name came from: `key` (the value that picked the
    /// case, read as a name: an enum's name, a chunk's four letters), `case`
    /// (the type of the case a switch took), `name` (the case is not a record,
    /// so the element's own name, as the listing labels it: an ELF section's
    /// `.rodata`), `type` (a list whose elements are all one type), `none`,
    /// and `other` for every group past the first [`GROUP_CAP`].
    pub group_from: &'static str,
    /// `content`, `machinery`, `padding`, `framing` or `gap`.
    pub role: &'static str,
    pub bits: u64,
    /// How many fields, or for a gap how many separate stretches.
    pub count: u64,
    /// For `gap` and `padding`: how many of the bits are in bytes that are
    /// zero, and how many were not looked at (see `Ledger::done`). The rest
    /// are in bytes that hold something.
    pub zero_bits: u64,
    pub unscanned_bits: u64,
    /// The first field or stretch counted here, to go to.
    pub first_path: Vec<usize>,
    pub first_offset_bits: u64,
    /// For `padding`: the alignment in bytes. 0 for everything else.
    pub align: u32,
}

/// The whole ledger so far.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Ledger {
    /// Largest first.
    pub rows: Vec<LedgerRow>,
    /// The space's length, which the rows are shares of.
    pub file_bits: u64,
    /// Bits the rows add up to. More than `file_bits` only where the template
    /// reads one stretch twice in a way nothing marks as a second reading.
    pub counted_bits: u64,
    /// True once every byte is in a row and every gap has been read.
    pub done: bool,
}

/// A row's key: part, group, role, alignment.
type RowKey = (u32, u32, &'static str, u32);

#[derive(Debug, Clone, Default)]
struct RowCount {
    bits: u64,
    count: u64,
    zero_bits: u64,
    unscanned_bits: u64,
    first_path: Vec<usize>,
    first_offset: u64,
}

/// A stretch whose bytes are still to be read to split it into zero and not,
/// once the walk is over.
#[derive(Debug, Clone)]
struct Stretch {
    from: u64,
    to: u64,
    scale: u64,
    row: RowKey,
    /// When the frame the stretch was found in was opened, in the walk's
    /// order: what tells a field placed over the stretch from one the stretch
    /// is inside. See `Placed`. None for padding, which is a field and never
    /// has anything placed over it.
    frame: Option<u64>,
    path: Vec<usize>,
}

/// A field that an offset put somewhere, and when the walk opened and closed
/// it: a gap is inside it when the gap's frame opened between the two.
#[derive(Debug, Clone, Copy)]
struct Placed {
    from: u64,
    to: u64,
    open: u64,
    close: u64,
}

/// What the ledger has found, kept between goes.
#[derive(Debug, Default)]
pub(super) struct Tally {
    rows: BTreeMap<RowKey, RowCount>,
    parts: Vec<(Vec<usize>, String)>,
    groups: Vec<(String, &'static str)>,
    group_index: rustc_hash::FxHashMap<(String, &'static str), u32>,
    stretches: Vec<Stretch>,
    placed: Vec<Placed>,
    /// Set once the gaps have had what was placed over them taken out.
    settled: bool,
    /// How far the scan of `stretches` has got: which one, and how far in.
    scan_at: (usize, u64),
    /// Bits read so far for the zero split, against `SCAN_CAP`.
    scanned: u64,
}

/// How much gap and padding the ledger reads to split it into zero and
/// nonzero bytes. A template that describes a header and nothing else leaves
/// the whole file a gap, and reading a few gigabytes to say how much of it is
/// zero is a question for the overview's byte classes, which read in the
/// background already. Past this the bits are counted as not looked at.
const SCAN_CAP: u64 = 256 * 1024 * 1024 * 8;

/// How many groups the ledger keeps apart. A variant is a name the file
/// chose, and a file that names every record differently would make a ledger
/// as long as the file.
pub const GROUP_CAP: usize = 256;

/// How much one step of the scan reads at a time.
const SCAN_CHUNK: u64 = 64 * 1024 * 8;

impl Tally {
    /// The part a path names, added the first time it is met.
    pub(super) fn part(&mut self, path: &[usize], name: &str) -> u32 {
        if let Some(i) = self.parts.iter().position(|(p, _)| p == path) {
            return i as u32;
        }
        self.parts.push((path.to_vec(), name.to_string()));
        (self.parts.len() - 1) as u32
    }

    /// The group a variant name names. Group 0 is none. Past [`GROUP_CAP`]
    /// groups, every new name goes in one group called `other`.
    pub(super) fn group(&mut self, name: &str, from: &'static str) -> u32 {
        if self.groups.is_empty() {
            self.groups.push((String::new(), "none"));
        }
        if let Some(i) = self.group_index.get(&(name.to_string(), from)) {
            return *i;
        }
        let (name, from) = if self.groups.len() >= GROUP_CAP { (String::new(), "other") } else { (name.to_string(), from) };
        if let Some(i) = self.group_index.get(&(name.clone(), from)) {
            return *i;
        }
        self.groups.push((name.clone(), from));
        let i = (self.groups.len() - 1) as u32;
        self.group_index.insert((name, from), i);
        i
    }

    /// A field covering `bits` bits, counted `count` times over.
    pub(super) fn field(&mut self, part: u32, group: u32, role: &'static str, align: u32, bits: u64, count: u64, path: &[usize], offset: u64) {
        let e = self.rows.entry((part, group, role, align)).or_default();
        if e.count == 0 {
            e.first_path = path.to_vec();
            e.first_offset = offset;
        }
        e.bits = e.bits.saturating_add(bits);
        e.count = e.count.saturating_add(count);
    }

    /// Padding from `from` to `to`, to be read for its zero bytes later.
    pub(super) fn padding(&mut self, part: u32, group: u32, align: u32, from: u64, to: u64, scale: u64, path: &[usize]) {
        let bits = (to - from).saturating_mul(scale);
        self.field(part, group, "padding", align, bits, scale, path, from);
        if to > from {
            self.stretches.push(Stretch { from, to, scale, row: (part, group, "padding", align), frame: None, path: path.to_vec() });
        }
    }

    /// Bits from `from` to `to` that no field of the frame opened at `frame`
    /// covers. Not counted yet: what was placed over it comes out first.
    pub(super) fn gap(&mut self, part: u32, group: u32, from: u64, to: u64, scale: u64, frame: u64, path: &[usize]) {
        if to > from {
            self.stretches.push(Stretch { from, to, scale, row: (part, group, "gap", 0), frame: Some(frame), path: path.to_vec() });
        }
    }

    /// A field an offset placed, from `from` to `to`, opened at `open`. Its
    /// `close` is filled in when it closes; see [`Tally::closed`].
    pub(super) fn placed(&mut self, from: u64, to: u64, open: u64) -> usize {
        self.placed.push(Placed { from, to, open, close: open });
        self.placed.len() - 1
    }

    pub(super) fn closed(&mut self, placed: usize, close: u64) {
        if let Some(p) = self.placed.get_mut(placed) {
            p.close = close;
        }
    }

    /// Take out of every gap the fields placed over it, and count what is
    /// left. Once, when the walk is over.
    fn settle(&mut self) {
        if self.settled {
            return;
        }
        self.settled = true;
        let mut placed = std::mem::take(&mut self.placed);
        placed.sort_by_key(|p| (p.from, p.to));
        let stretches = std::mem::take(&mut self.stretches);
        let mut kept = Vec::with_capacity(stretches.len());
        for s in stretches {
            let Some(frame) = s.frame else {
                kept.push(s);
                continue;
            };
            // What was placed over this gap: everything that starts inside
            // it, and a few before it that may reach into it. A field whose
            // own structure the gap is in is not taken out, since the gap is
            // bytes that field left.
            let lo = placed.partition_point(|p| p.from < s.from);
            let hi = placed.partition_point(|p| p.from < s.to);
            let mut over: Vec<(u64, u64)> = placed[lo.saturating_sub(64)..hi]
                .iter()
                .filter(|p| p.to > s.from && p.from < s.to && !(p.open <= frame && frame <= p.close))
                .map(|p| (p.from.max(s.from), p.to.min(s.to)))
                .collect();
            over.sort_unstable();
            let mut at = s.from;
            let mut pieces = Vec::new();
            for (a, b) in over {
                if a > at {
                    pieces.push((at, a));
                }
                at = at.max(b);
            }
            if at < s.to {
                pieces.push((at, s.to));
            }
            for (from, to) in pieces {
                let e = self.rows.entry(s.row).or_default();
                if e.count == 0 {
                    e.first_path = s.path.clone();
                    e.first_offset = from;
                }
                e.bits = e.bits.saturating_add((to - from).saturating_mul(s.scale));
                e.count = e.count.saturating_add(s.scale);
                kept.push(Stretch { from, to, ..s.clone() });
            }
        }
        // Everything starts out not looked at, and comes off as it is read.
        for s in &kept {
            if let Some(e) = self.rows.get_mut(&s.row) {
                e.unscanned_bits = e.unscanned_bits.saturating_add((s.to - s.from).saturating_mul(s.scale));
            }
        }
        self.stretches = kept;
    }

    /// Whether every gap has been read, or given up on.
    pub(super) fn scanned(&self) -> bool {
        self.settled && self.scan_at.0 >= self.stretches.len()
    }

    /// Read the gaps and padding for zero bytes, a chunk at a time, charging
    /// each chunk against the allowance. Only in the walk's own space.
    pub(super) fn scan<S: Source>(&mut self, ev: &mut Evaluator, doc: &Document<S>, space: u32) -> R<()> {
        self.settle();
        while let Some(s) = self.stretches.get(self.scan_at.0) {
            let s = s.clone();
            let at = s.from.max(self.scan_at.1);
            if at >= s.to || self.scanned >= SCAN_CAP {
                self.scan_at = (self.scan_at.0 + 1, 0);
                continue;
            }
            // Whole bytes only: a gap that starts or ends inside a byte has its
            // odd bits left as not looked at.
            let from = at.div_ceil(8) * 8;
            let to = (s.to / 8 * 8).min(from + SCAN_CHUNK);
            if to <= from {
                self.scan_at = (self.scan_at.0 + 1, 0);
                continue;
            }
            ev.spend(from)?;
            let bytes = ev.read_in(doc, space, from, to - from)?;
            let zeros = bytes.iter().filter(|b| **b == 0).count() as u64 * 8;
            let read = to - from;
            if let Some(e) = self.rows.get_mut(&s.row) {
                e.zero_bits = e.zero_bits.saturating_add(zeros.saturating_mul(s.scale));
                e.unscanned_bits = e.unscanned_bits.saturating_sub(read.saturating_mul(s.scale));
            }
            self.scanned = self.scanned.saturating_add(read);
            self.scan_at = if to >= s.to / 8 * 8 { (self.scan_at.0 + 1, 0) } else { (self.scan_at.0, to) };
        }
        Ok(())
    }

    pub(super) fn ledger(&self, file_bits: u64, done: bool) -> Ledger {
        let mut rows: Vec<LedgerRow> = self
            .rows
            .iter()
            // A field of no bits is a field and not a stretch of the file:
            // nothing to put in a ledger of where the bits go.
            .filter(|(_, c)| c.bits > 0)
            .map(|(&(part, group, role, align), c)| {
                let (part_path, part_name) = self.parts.get(part as usize).cloned().unwrap_or_default();
                let (group, group_from) = self.groups.get(group as usize).cloned().unwrap_or((String::new(), "none"));
                LedgerRow {
                    part: part_path,
                    part_name,
                    group,
                    group_from,
                    role,
                    bits: c.bits,
                    count: c.count,
                    zero_bits: c.zero_bits,
                    unscanned_bits: c.unscanned_bits,
                    first_path: c.first_path.clone(),
                    first_offset_bits: c.first_offset,
                    align,
                }
            })
            .collect();
        rows.sort_by(|a, b| b.bits.cmp(&a.bits).then_with(|| (&a.part, &a.group, a.role).cmp(&(&b.part, &b.group, b.role))));
        let counted_bits = rows.iter().map(|r| r.bits).fold(0u64, u64::saturating_add);
        Ledger { rows, file_bits, counted_bits, done }
    }
}

/// A list element's name with its index taken off: `[3] IDAT` is `IDAT`.
/// Nothing when the name is only the index.
pub(super) fn without_index(label: &str) -> Option<&str> {
    let rest = match label.strip_prefix('[') {
        Some(r) => match r.find(']') {
            Some(i) if r[..i].bytes().all(|b| b.is_ascii_digit()) => r[i + 1..].trim_start(),
            _ => label,
        },
        None => label,
    };
    (!rest.is_empty()).then_some(rest)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_name_loses_its_index() {
        assert_eq!(without_index("[3] IDAT"), Some("IDAT"));
        assert_eq!(without_index("[12]"), None);
        assert_eq!(without_index(".text"), Some(".text"));
        assert_eq!(without_index("[a] b"), Some("[a] b"));
    }

    #[test]
    fn a_gap_loses_what_was_placed_over_it_but_not_what_it_is_inside() {
        let mut t = Tally::default();
        let part = t.part(&[], "root");
        let g = t.group("", "none");
        // The root's frame, opened at 1, leaves 0..800 as a gap; a field placed
        // at 100..300, opened at 5, covers some of it; a frame inside that
        // field, opened at 6, leaves 200..250 as a gap of its own.
        t.gap(part, g, 0, 800, 1, 1, &[]);
        let p = t.placed(100, 300, 5);
        t.gap(part, g, 200, 250, 1, 6, &[0]);
        t.closed(p, 9);
        t.settle();
        let l = t.ledger(800, true);
        let gap = l.rows.iter().find(|r| r.role == "gap").expect("a gap row");
        assert_eq!(gap.bits, 600 + 50, "0..100 and 300..800 of the root's, and the field's own 200..250");
        assert_eq!(gap.count, 3);
    }
}

File: web/src/report/ledger.ts
// The core's byte ledger, read for the report: every bit of the file in one
// row, keyed by part, by the variant of the nearest list element, and by role.
// See "The byte ledger" in docs/DESIGN-report-view.md.
//
// The report shows it twice. Section 6 has one row per part and group, with
// its share of the file; section 12 has every role under each of those. The
// grouping here is the part that can be wrong, so it has no DOM in it and the
// tests run it.

import type { TemplateNode } from "../doc.ts";
import type { LedgerRole, LedgerRow } from "./coredata.ts";
import { stripIndex } from "./partrules.ts";

/** One line of the ledger as the report shows it: a part, or one group of the
 *  elements of a list inside a part. */
export type LedgerLine = {
  readonly key: string;
  /** The part's name, or null for the fields of a structure taken together. */
  readonly part: string | null;
  /** What the fields taken together belong to, for a line of plain fields. */
  readonly parentPath: readonly number[] | null;
  /** The group's name, "" where there is none. */
  readonly group: string;
  /** True when the group is named by a type in the template rather than by a
   *  value in the file: `TableLeaf`, which the report writes as words. */
  readonly byType: boolean;
  /** The part the group is in, and whether the report names it after the
   *  group: only where the grouped lines are in more than one part. A group
   *  in the one list the file is, or in the only list with groups, is named
   *  by the group alone, since "in <list>" would say nothing. */
  readonly partPath: readonly number[];
  readonly partShown: boolean;
  /** The names of the fields taken together, for a line of plain fields. */
  readonly fields: readonly string[];
  readonly bits: number;
  readonly firstOffsetBits: number;
  readonly firstPath: readonly number[];
  /** The line's bits by role, largest first. */
  readonly roles: readonly RoleShare[];
};

export type RoleShare = {
  readonly role: LedgerRole;
  readonly bits: number;
  /** For padding and gaps: bits in zero bytes, and bits not read. */
  readonly zeroBits: number;
  readonly unscannedBits: number;
  /** For padding: the alignments it pads to. */
  readonly aligns: readonly number[];
};

const samePath = (a: readonly number[], b: readonly number[]): boolean => a.length === b.length && a.every((x, i) => x === b[i]);

/**
 * The ledger's rows as the report's lines, in file order.
 *
 * - A part with groups under it is a line per group: a ZIP's records are the
 *   local files, the central directory and its end.
 * - A part that is one plain field is taken together with the other plain
 *   fields of the same structure: the thirty fields of an ELF header are one
 *   line, not thirty. One such field on its own keeps its own name.
 * - Bytes no field describes are one line, and padding another.
 */
export function ledgerLines(rows: readonly LedgerRow[]): LedgerLine[] {
  type Acc = {
    key: string;
    part: string | null;
    parentPath: readonly number[] | null;
    group: string;
    fields: string[];
    bits: number;
    firstOffsetBits: number;
    firstPath: readonly number[];
    partPath: readonly number[];
    /** The group is named by the structure's type, not by a value. */
    byType: boolean;
    /** The group's name is a type's, whether the list's one type or the
     *  type of the case a switch took. */
    typeNamed: boolean;
    roles: Map<LedgerRole, RoleAcc>;
  };
  const lines = new Map<string, Acc>();
  for (const r of rows) {
    const plain = r.group === "" && r.role !== "gap" && r.role !== "padding" && samePath(r.first_path, r.part);
    const key =
      r.role === "gap"
        ? "gap"
        : r.role === "padding"
          ? "padding"
          : plain
            ? `fields:${r.part.slice(0, -1).join("/")}`
            : `part:${r.part.join("/")}\u0000${r.group}`;
    let a = lines.get(key);
    if (a === undefined) {
      a = {
        key,
        part: plain || r.role === "gap" || r.role === "padding" ? null : r.part_name,
        parentPath: plain ? r.part.slice(0, -1) : null,
        group: plain ? "" : r.group,
        fields: [],
        bits: 0,
        firstOffsetBits: r.first_offset_bits,
        firstPath: r.first_path,
        partPath: r.part,
        byType: r.group_from === "type",
        typeNamed: r.group_from === "type" || r.group_from === "case",
        roles: new Map(),
      };
      lines.set(key, a);
    }
    if (plain && !a.fields.includes(r.part_name)) a.fields.push(r.part_name);
    a.bits += r.bits;
    if (r.first_offset_bits < a.firstOffsetBits) {
      a.firstOffsetBits = r.first_offset_bits;
      a.firstPath = r.first_path;
    }
    const role: RoleAcc = a.roles.get(r.role) ?? { bits: 0, zero: 0, unscanned: 0, aligns: new Set<number>() };
    role.bits += r.bits;
    role.zero += r.zero_bits;
    role.unscanned += r.unscanned_bits;
    if (r.align > 0) role.aligns.add(r.align);
    a.roles.set(r.role, role);
  }
  // A structure inside another group's element that is named only by its
  // type is part of that group's line: a JPEG dht segment's Huffman tables,
  // a ZIP local file's extra fields. The line then says how much of the file
  // the segments take. The inner group's first field lies in the element
  // where the outer group's first field is, further in. A group named by a
  // value in the file stays a line of its own, since the value says
  // something: a MIDI track's note-on events, a JPEG table's precision.
  const owners = new Map<string, Acc>();
  const nested: Acc[] = [];
  const byDepth = [...lines.values()].filter((a) => a.key.startsWith("part:")).sort((x, y) => x.firstPath.length - y.firstPath.length);
  for (const a of byDepth) {
    const element = `${a.partPath.join("/")}|${a.firstPath.slice(0, a.partPath.length + 1).join("/")}`;
    const owner = owners.get(element);
    if (owner === undefined) owners.set(element, a);
    else if (a.byType && owner.firstPath.length < a.firstPath.length) {
      mergeInto(owner, a);
      nested.push(a);
    }
  }
  for (const a of nested) lines.delete(a.key);
  // The parts that hold groups. Only where there are two or more does a
  // group's line say which part it is in: a MIDI file's events are all in the
  // one list the file is, and "meta in file" would read as "in the file".
  const grouped = new Set([...lines.values()].filter((a) => a.group !== "").map((a) => a.partPath.join("/")));
  return [...lines.values()]
    .map((a) => ({
      key: a.key,
      // One plain field on its own is that field.
      part: a.part ?? (a.fields.length === 1 ? (a.fields[0] ?? null) : null),
      parentPath: a.fields.length === 1 ? null : a.parentPath,
      group: a.group,
      byType: a.group !== "" && a.typeNamed,
      partPath: a.partPath,
      partShown: a.group !== "" && a.partPath.length > 0 && grouped.size > 1,
      fields: a.fields,
      bits: a.bits,
      firstOffsetBits: a.firstOffsetBits,
      firstPath: a.firstPath,
      roles: [...a.roles.entries()]
        .map(([role, v]) => ({ role, bits: v.bits, zeroBits: v.zero, unscannedBits: v.unscanned, aligns: [...v.aligns].sort((x, y) => x - y) }))
        .sort((x, y) => y.bits - x.bits),
    }))
    .sort((x, y) => x.firstOffsetBits - y.firstOffsetBits);
}

type RoleAcc = { bits: number; zero: number; unscanned: number; aligns: Set<number> };
type LineAcc = { bits: number; firstOffsetBits: number; firstPath: readonly number[]; roles: Map<LedgerRole, RoleAcc> };

function mergeInto(owner: LineAcc, a: LineAcc): void {
  owner.bits += a.bits;
  if (a.firstOffsetBits < owner.firstOffsetBits) {
    owner.firstOffsetBits = a.firstOffsetBits;
    owner.firstPath = a.firstPath;
  }
  for (const [role, v] of a.roles) {
    const o = owner.roles.get(role) ?? { bits: 0, zero: 0, unscanned: 0, aligns: new Set<number>() };
    o.bits += v.bits;
    o.zero += v.zero;
    o.unscanned += v.unscanned;
    for (const x of v.aligns) o.aligns.add(x);
    owner.roles.set(role, o);
  }
}

/** True for the line of bytes no field describes. */
export function isGapLine(l: LedgerLine): boolean {
  return l.key === "gap";
}

/**
 * A part's fields as rows: one that covers the whole part would repeat it and
 * is left out, and neighbours of the same name are one row with a count, so
 * four dht segments in a row are `dht × 4`.
 */
export function fieldRows(u: { offsetBits: number; sizeBits: number }, kids: readonly Pick<TemplateNode, "name" | "path" | "offset_bits" | "size_bits">[]): { name: string; path: readonly number[]; startBit: number; endBit: number; count: number }[] {
  const out: { name: string; path: readonly number[]; startBit: number; endBit: number; count: number }[] = [];
  for (const k of kids) {
    if (k.offset_bits === u.offsetBits && k.size_bits === u.sizeBits) continue;
    const name = stripIndex(k.name) || k.name;
    const last = out[out.length - 1];
    if (last !== undefined && last.name === name) {
      last.endBit = Math.max(last.endBit, k.offset_bits + k.size_bits);
      last.count++;
      continue;
    }
    out.push({ name, path: k.path, startBit: k.offset_bits, endBit: k.offset_bits + k.size_bits, count: 1 });
  }
  return out;
}
